cmd/keys.go

```diff
@@ -131,61 +132,111 @@ func keysRestoreCmd(a *appState) *cobra.Command {
 		Use:     "restore chain_name key_name mnemonic",
 		Aliases: []string{"r"},
 		Short:   "Restores a mnemonic to the keychain associated with a particular chain",
-		Args:    withUsage(cobra.ExactArgs(3)),
+		Args:    withUsage(cobra.RangeArgs(2, 3)),
 		Example: strings.TrimSpace(fmt.Sprintf(`
-$ %s keys restore ibc-0 testkey "[mnemonic-words]"
-$ %s k r cosmoshub faucet-key "[mnemonic-words]"`, appName, appName)),
+$ %s keys restore  ibc-0 testkey "[mnemonic-words]"
+$ %s k r "[mnemonic-words]" cosmoshub faucet-key
+$ %s k r demo-key "[mnemonic-words]" --restore-all`, appName, appName, appName)),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			keyName := args[1]
-
-			chain, ok := a.config.Chains[args[0]]
-			if !ok {
-				return errChainNotFound(args[0])
-			}
-
-			if chain.ChainProvider.KeyExists(keyName) {
-				return errKeyExists(keyName)
-			}
 
 			cmdFlags := cmd.Flags()
 
-			coinType, err := cmdFlags.GetInt32(flagCoinType)
+			restoreAll, err := cmdFlags.GetBool(flagRestoreAll)
 			if err != nil {
 				return err
 			}
 
-			if coinType < 0 {
-				if ccp, ok := chain.ChainProvider.(*cosmos.CosmosProvider); ok && ccp.PCfg.Slip44 != nil {
-					coinType = int32(*ccp.PCfg.Slip44)
-				} else {
-					coinType = int32(defaultCoinType)
+			if !restoreAll {
+
+				keyName := args[1]
+				chain, ok := a.config.Chains[args[0]]
+				if !ok {
+					return errChainNotFound(args[0])
 				}
-			}
 
-			algo, err := cmdFlags.GetString(flagAlgo)
-			if err != nil {
-				return err
-			}
+				if chain.ChainProvider.KeyExists(keyName) {
+					return errKeyExists(keyName)
+				}
 
-			if algo == "" {
-				if ccp, ok := chain.ChainProvider.(*cosmos.CosmosProvider); ok {
-					algo = ccp.PCfg.SigningAlgorithm
-				} else {
-					algo = string(hd.Secp256k1Type)
+				coinType, err := cmdFlags.GetInt32(flagCoinType)
+				if err != nil {
+					return err
 				}
-			}
 
-			address, err := chain.ChainProvider.RestoreKey(keyName, args[2], uint32(coinType), algo)
-			if err != nil {
-				return err
-			}
+				if coinType < 0 {
+					if ccp, ok := chain.ChainProvider.(*cosmos.CosmosProvider); ok && ccp.PCfg.Slip44 != nil {
+						coinType = int32(*ccp.PCfg.Slip44)
+					} else {
+						coinType = int32(defaultCoinType)
+					}
+				}
 
-			fmt.Fprintln(cmd.OutOrStdout(), address)
+				algo, err := cmdFlags.GetString(flagAlgo)
+				if err != nil {
+					return err
+				}
+
+				if algo == "" {
+					if ccp, ok := chain.ChainProvider.(*cosmos.CosmosProvider); ok {
+						algo = ccp.PCfg.SigningAlgorithm
+					} else {
+						algo = string(hd.Secp256k1Type)
+					}
+				}
+
+				address, err := chain.ChainProvider.RestoreKey(keyName, args[2], uint32(coinType), algo)
+				if err != nil {
+					return err
+				}
+				fmt.Fprintln(cmd.OutOrStdout(), address)
+				return nil
+
+			} else {
+
+				chains := a.config.Chains
+				keyName := args[0]
+				for _, c := range chains {
+
```

Comment:
we probably still want to check if the key already exists like we do in the case where we are restoring just a single mnemonic

```go
if chain.ChainProvider.KeyExists(keyName) {
    return errKeyExists(keyName)
}
```